src/persist-client/src/internal/cache.rs

```diff
@@ -0,0 +1,691 @@
+
+use std::sync::{Arc, Mutex};
+
+use async_trait::async_trait;
+use bytes::Bytes;
+use mz_ore::bytes::SegmentedBytes;
+use mz_ore::cast::CastFrom;
+use mz_persist::location::{Atomicity, Blob, BlobMetadata, ExternalError};
+
+use crate::cfg::{DynamicConfig, PersistConfig};
+use crate::internal::metrics::Metrics;
+
+// In-memory cache for [Blob].
+#[derive(Debug)]
+pub struct BlobMemCache {
+    cfg: Arc<DynamicConfig>,
+    metrics: Arc<Metrics>,
+    cache: Mutex<lru::Lru<String, SegmentedBytes>>,
+    blob: Arc<dyn Blob + Send + Sync>,
+}
+
+impl BlobMemCache {
+    pub fn new(
+        cfg: &PersistConfig,
+        metrics: Arc<Metrics>,
+        blob: Arc<dyn Blob + Send + Sync>,
+    ) -> Arc<dyn Blob + Send + Sync> {
+        let eviction_metrics = Arc::clone(&metrics);
+        let cache = lru::Lru::new(cfg.dynamic.blob_cache_mem_limit_bytes(), move |_, _, _| {
+            eviction_metrics.blob_cache_mem.evictions.inc()
+        });
+        let blob = BlobMemCache {
+            cfg: Arc::clone(&cfg.dynamic),
+            metrics,
+            cache: Mutex::new(cache),
+            blob,
+        };
+        Arc::new(blob)
+    }
+
+    fn resize_and_update_size_metrics(&self, cache: &mut lru::Lru<String, SegmentedBytes>) {
+        cache.update_capacity(self.cfg.blob_cache_mem_limit_bytes());
+        self.metrics
+            .blob_cache_mem
+            .size_blobs
+            .set(u64::cast_from(cache.entry_count()));
+        self.metrics
+            .blob_cache_mem
+            .size_bytes
+            .set(u64::cast_from(cache.entry_weight()));
+    }
+}
+
+#[async_trait]
+impl Blob for BlobMemCache {
+    async fn get(&self, key: &str) -> Result<Option<SegmentedBytes>, ExternalError> {
+        // First check if the blob is in the cache. If it is, return it. If not,
+        // fetch it and put it in the cache.
+        //
+        // Blobs are write-once modify-never, so we don't have to worry about
+        // any races or cache invalidations here. If the value is in the cache,
+        // it's also what's in s3 (if not, then there's a horrible bug somewhere
+        // else).
+        if let Some((_, cached_value)) = self.cache.lock().expect("lock poisoned").get(key) {
+            self.metrics.blob_cache_mem.hits_blobs.inc();
+            self.metrics
+                .blob_cache_mem
+                .hits_bytes
+                .inc_by(u64::cast_from(cached_value.len()));
+            return Ok(Some(cached_value.clone()));
+        }
+
+        let res = self.blob.get(key).await?;
+        if let Some(blob) = res.as_ref() {
+            // TODO: It would likely be useful to allow a caller to opt out of
+            // adding the data to the cache (e.g. compaction inputs, perhaps
+            // some read handles).
+            let mut cache = self.cache.lock().expect("lock poisoned");
+            cache.insert(key.to_owned(), blob.clone(), blob.len());
+            self.resize_and_update_size_metrics(&mut cache);
```

Comment:
IIUC this will end up evicting from the underlying map twice: once for the insert  and once on the resize. Seems harmless... but it may be cleaner to not resize during map updates and do all the evicting on resize, since it's called for every update anyways.

Reply:
IMO it's important for `insert` to hold all the invariants at the time it returns. And then resize is a conceptually separate operation, I think it's an implementation detail of `BlobMemCache` that it happens to call it each time the map is updated